### PR TITLE
Make Root module fully differentiable

### PR DESCRIPTION
**Pull request by SCiarella**
This PR addresses the differentiability requirement of issue #47.

All the parameters except `RDRRTB` were already differentiable since we have defined them as `torch.tensor`. 
To make  `RDRRTB` differentiable, I had to redefine `pcse.utils.Afgen` to use tensors instead of `list`. For this reason I have added `src/physical_models/crop/afgen.py` which is now differentiable.

I have also restructured  `test_root_dynamics` such that all the combinations of parameter-output are checked, and code duplication is reduced.

-----
As a quality of life improvement, I have installed [`pre-commit`](https://pre-commit.com/) that automatically runs `ruff` checks before each `git commit`.

### CODE REVIEW

File: tests/physical_models/crop/test_root_dynamics.py [hunk 1]
```diff
@@ -147,53 +147,109 @@ def test_wofost_pp_with_root_dynamics(self):
...
+
+# Define which parameter-output pairs should have gradients
+# Format: {param_name: [list of outputs that should have gradients]}
+GRADIENT_MAPPING = {
+    "TDWI": ["TWRT"],  # e.g. TDWI affects TWRT so I put it here
+    "RDI": ["RD"],
+    "RRI": ["RD"],
+    "RDMCR": ["RD"],
+    "RDMSOL": ["RD"],
+    "RDRRTB": ["TWRT"],
```
Comment (SarahAlidoost) [inline tests/physical_models/crop/test_root_dynamics.py]: This list provides useful information for users when calculating gradients for example in optimization. Can we add a comments about this in the docstring of root_dynamics? thanks!

File: tests/physical_models/crop/test_root_dynamics.py [hunk 2]
```diff
@@ -147,53 +147,109 @@ def test_wofost_pp_with_root_dynamics(self):
...
 
-class TestDiffRootDynamicsTDWI:
-    def test_gradients_tdwi_rd_root_dynamics(self):
+# Define parameters and outputs
+PARAM_NAMES = ["TDWI", "RDI", "RRI", "RDMCR", "RDMSOL", "IAIRDU", "RDRRTB"]
+OUTPUT_NAMES = ["RD", "TWRT"]
+
+# Define parameter configurations (value, dtype)
+PARAM_DEFAULT_VAL = {
+    "TDWI": (0.2, torch.float32),
```
Comment (SarahAlidoost) [inline tests/physical_models/crop/test_root_dynamics.py]: One of the requirements in #39 is to vectorize the operations in the module. As an example, see [leaf_dynamics](https://github.com/WUR-AI/diffWOFOST/blob/main/src/diffwofost/physical_models/crop/leaf_dynamics.py) and [its tests regarding model correctness](https://github.com/WUR-AI/diffWOFOST/blob/02452f5b136587e66369f044b69bdc515cabbd8b/tests/physical_models/crop/test_leaf_dynamics.py#L122-L320) as well as [its tests regarding gradients](https://github.com/WUR-AI/diffWOFOST/blob/02452f5b136587e66369f044b69bdc515cabbd8b/tests/physical_models/crop/test_leaf_dynamics.py#L355-L356). We want to implement the same ideas for `root_dynamics`, making sure all computations are vectorized.

File: src/diffwofost/physical_models/afgen.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    default_value = Afgen([0, 0, 1, 1])
+    into_text = "An AFGEN table of XY pairs"
+
+    def validate(self, obj, value):
+        """Validate that the value is an Afgen instance or an iterable to create one."""
+        if isinstance(value, Afgen):
+            return value
+        elif isinstance(value, Iterable):
+            return Afgen(value)
+        self.error(obj, value)
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/diffwofost/physical_models/afgen.py]: Sure! I have now merged afgen into utils and updated the docstrings

File: tests/physical_models/crop/test_root_dynamics.py (hunk 2 above)
Comment (SCiarella) [reply tests/physical_models/crop/test_root_dynamics.py]: Sorry, I do not fully understand this point, can you give me more details or show me an example of desired behavior?

File: tests/physical_models/crop/test_root_dynamics.py [hunk 4]
```diff
@@ -147,53 +147,109 @@ def test_wofost_pp_with_root_dynamics(self):
...
+    "RDI": (0.2, torch.float32),
+    "RRI": (0.2, torch.float32),
+    "RDMCR": (0.2, torch.float32),
+    "RDMSOL": (0.2, torch.float32),
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply tests/physical_models/crop/test_root_dynamics.py]: Thanks!

File: tests/physical_models/crop/test_root_dynamics.py [hunk 5]
```diff
@@ -147,53 +147,109 @@ def test_wofost_pp_with_root_dynamics(self):
...
+# Parametrize decorators for gradient tests
+gradient_test_params = pytest.mark.parametrize(
+    "param_name,output_name",
+    _gradient_params,
+)
+no_gradient_test_params = pytest.mark.parametrize(
+    "param_name,output_name",
+    _no_gradient_params,
+)
+
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply tests/physical_models/crop/test_root_dynamics.py]: fixed

File: tests/physical_models/crop/test_root_dynamics.py [hunk 6]
```diff
@@ -119,6 +123,205 @@ def test_root_dynamics_with_engine(self):
                 config_path,
             )
 
+    @pytest.mark.parametrize("param", ["RDI", "RRI", "RDMCR", "RDMSOL", "TDWI", "IAIRDU"])
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply tests/physical_models/crop/test_root_dynamics.py]: The parameter "RDRRTB" should be also in the list, right? I think the root_dynamic needs modification to handle "RDRRTB" when its value is ndarray something like `[[0.0, 0.0, 1.5, 0.02], [0.0, 0.0, 1.6, 0.03]]`.

File: tests/physical_models/crop/test_root_dynamics.py [hunk 7]
```diff
@@ -119,6 +123,205 @@ def test_root_dynamics_with_engine(self):
...
+        (
+            crop_model_params_provider,
+            weather_data_provider,
+            agro_management_inputs,
+            external_states,
+        ) = prepare_engine_input(test_data_path, crop_model_params)
+        config_path = str(phy_data_folder / "WOFOST_Root_Dynamics.conf")
+
+        # Setting a vector (with one value) for the RDI and RRI parameters
+        for param in ("RDI", "RRI"):
```
Comment (SarahAlidoost) [inline tests/physical_models/crop/test_root_dynamics.py]: why not to test all the parameters?

**Review by SarahAlidoost — CHANGES_REQUESTED** [review]
@SCiarella thanks for addressing the comments 👍  We need a few more tests to make sure everything works with vectorization:  
1. Tests for gradients when parameters are vectors, see my code suggestions in class `TestDiffRootDynamicsGradients`. 
2. Tests for parameter "RDRRTB", see the list in `test_root_dynamics_with_one_parameter_vector`. 
3. Tests for other parameters in  `test_root_dynamics_with_multiple_parameter_vectors`, and `test_root_dynamics_with_multiple_parameter_arrays`. Currently, only two parameters ("RDI", "RRI") are there. 

Please have a look and let me know if something is unclear. Thanks.

File: tests/physical_models/crop/test_root_dynamics.py [hunk 8]
```diff
@@ -147,53 +350,104 @@ def test_wofost_pp_with_root_dynamics(self):
...
+
+    # Define parameter configurations (value, dtype)
+    param_default_val = {
+        "TDWI": (0.2, torch.float32),
+        "RDI": (10, torch.float32),
+        "RRI": (1.0, torch.float32),
+        "RDMCR": (120, torch.float32),
+        "RDMSOL": (120, torch.float32),
+        "IAIRDU": (0.2, torch.float32),
+        "RDRRTB": ([0.0, 0.0, 1.5, 0.02], torch.float32),  # Table parameter: [x1, y1, x2, y2]
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply tests/physical_models/crop/test_root_dynamics.py]: You are right! I have done this in the last commit

File: tests/physical_models/crop/test_root_dynamics.py [hunk 9]
```diff
@@ -119,6 +123,205 @@ def test_root_dynamics_with_engine(self):
...
+            crop_model_params_provider,
+            weather_data_provider,
+            agro_management_inputs,
+            external_states,
+        ) = prepare_engine_input(test_data_path, crop_model_params)
+        config_path = str(phy_data_folder / "WOFOST_Root_Dynamics.conf")
+
+        # Setting an array with arbitrary shape (and one value) for the
+        # RDI and RRI parameters
+        for param in ("RDI", "RRI"):
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply tests/physical_models/crop/test_root_dynamics.py]: yes

**Comment by SCiarella** [comment]
Thanks @SarahAlidoost! 

Now I have fully vectorized also `RDRRTB` by extending the `Afgen` class to work with tensors of tables. I have extended the unit tests in `test_utils` to also cover the tensor case and I have expanded the `test_root_dynamics` to cover the case when all the parameters (including `RDRRTB`) are tensors, as you pointed out in the code review.

Notice that I have also added a branch to `calculate_numerical_grad` in order to return the correct gradient for `RDRRTB`.

File: src/diffwofost/physical_models/utils.py [hunk 10]
```diff
@@ -289,15 +293,342 @@ def get_test_data(file_path):
...
+
+        model = get_model_fn()
+        output = model({param_name: torch.nn.Parameter(p_plus)})
+        loss_plus = output[out_name].sum(dim=0)
+
+        model = get_model_fn()
+        output = model({param_name: torch.nn.Parameter(p_minus)})
+        loss_minus = output[out_name].sum(dim=0)
+
+        return (loss_plus.data - loss_minus.data) / (2 * delta)
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/utils.py]:
```suggestion
    # Parameters like RDRRTB are batched tables, so we need to compute
    # the gradient for each table element separately
    # So, we flatten the parameter for easier indexing
    param_flat = param_value.reshape(-1)
    grad_flat = torch.zeros_like(param_flat)

    for i in range(param_flat.numel()):
        p_plus = param_flat.clone()
        p_plus[i] += delta
        p_minus = param_flat.clone()
        p_minus[i] -= delta

        p_plus = p_plus.view_as(param_value)
        p_minus = p_minus.view_as(param_value)

        model = get_model_fn()
        out_plus = model({param_name: p_plus})[out_name]
        loss_plus = out_plus.sum()

        model = get_model_fn()
        out_minus = model({param_name: p_minus})[out_name]
        loss_minus = out_minus.sum()

        grad_flat[i] = (loss_plus - loss_minus) / (2 * delta)

    return grad_flat.view_as(param_value)
```

**Review by SarahAlidoost — APPROVED** [review]
@SCiarella thanks for addressing the comments, nice work 💪 I have one suggestion on refactoring of numerical function making it more generic, see my suggestion on the code.  If you agree with it, please go ahead with committing it and merge this PR 🚀
